File: python/SNNDPC/Classifier.py
from typing import Tuple

from numpy import ndarray

from .Core import Core
from .Option import Option

class Classifier(Core):
	def __init__(self, numNeighbor: int, numRecord: int, numDimension: int, numCentroid: int, option: Option = Option()) -> None:
		"""
		Initialize a SNN-DPC classifier.

		Args:
			numNeighbor: A.k.a. k, number of neighbors should be evaluated.
			numRecord: A.k.a. n, number of records in your dataset
			numDimension: A.k.a. d, number of dimensions in your dataset
			numCentroid: A.k.a. nc, number of centroids
			option: Other options, see the documentation of corresponding fields
		"""
		super().__init__(numNeighbor, numRecord, numDimension, numCentroid, option)

	def Run(self, data: ndarray) -> Tuple[ndarray, ndarray]:
		"""
		End-to-end execution.

		No more confusion on how to use this classifier, just call this and enjoy.

		Args:
			data: Input data of size [numRecord, numDimension].

		Returns:
			Indices of centroids, assignment of non-centroids.
		"""
		self.Input(data)
		self.ComputeDistance()
		self.ComputeNeighbor()
		self.ComputeSharedNeighbor()
		self.ComputeSimilarity()
		self.ComputeRho()
		self.ComputeDelta()
		self.ComputeGamma()
		self.ComputeCentroid()
		self.AssignNonCentroidStep1()
		self.AssignNonCentroidStep2()
		return self.Output()

	def Input(self, data: ndarray) -> None:
		"""
		Input the dataset.

		Args:
			data: Input data of size [numRecord, numDimension].
		"""
		if self.option.Preprocess:
			self.data: ndarray = self.option.Preprocess(data)
		elif self.option.shouldCopyInputData:
			self.data: ndarray = data.copy()
		else:
			self.data: ndarray = data

	def Output(self) -> Tuple[ndarray, ndarray]:
		"""
		Output the result.

		Returns:
			Indices of centroids and the assignment of non-centroids.
		"""

		if self.option.shouldUseOrdinalAssignment:
			return self.indexCentroid, self.indexCentroid[self.indexAssignment]
		else:
			return self.indexCentroid, self.indexAssignment

File: GPU/Classifier.cpp
#include "Classifier.h"

#include <cuda_runtime.h>

using namespace std;
using namespace SNNDPC::GPU;

template<typename T>
Classifier<T>::Classifier(int numNeighbor, int numRecord, int numDimension, int numCentroid, const SNNDPC::Option& option): Core<T>(numNeighbor, numRecord, numDimension, numCentroid, option) { }

template<typename T>
Classifier<T>::~Classifier() {
	cudaFree(this->data);
}

template<typename T>
void Classifier<T>::Run(const T* data, int* centroid, int* assignment) {
	Input(data);
	Normalize();
	ComputeDistance();
	ComputeNeighbor();
	ComputeSharedNeighbor();
	ComputeSimilarity();
	ComputeRho();
	ComputeDelta();
	ComputeGamma();
	ComputeCentroid();
	AssignNonCentroidStep1();
	AssignNonCentroidStep2();
	Output(centroid, assignment);
}

template<typename T>
void Classifier<T>::Input(const T* data) {
	cudaMallocManaged(&this->data, n * d * sizeof(T));
	cudaMemcpy(this->data, data, n * d * sizeof(T), cudaMemcpyHostToDevice);
}

template<typename T>
void Classifier<T>::Normalize() {
	for (int i = 0; i < d; i++) {
		int indexLeast, indexMost;
		cublasIaminEx(handle, n, data + i, type, d, &indexLeast);
		cublasIamaxEx(handle, n, data + i, type, d, &indexMost);
		const T* least = &data[(indexLeast - 1) * d + i];
		const T* most = &data[(indexMost - 1) * d + i];
		const T diff = T{1} / (*most - *least);
		const T MinusOne = -1;
		cublasAxpyEx(handle, n, &MinusOne, type, least, type, 0, data + i, type, d, type);
		cublasScalEx(handle, n, &diff, type, data + i, type, d, type);
	}
}

template<typename T>
void Classifier<T>::Output(int* centroid, int* assignment) {
	if (centroid) cudaMemcpy(centroid, indexCentroid, nc * sizeof(int), cudaMemcpyDeviceToHost);
	if (assignment) cudaMemcpy(assignment, indexAssignment, n * sizeof(int), cudaMemcpyDeviceToHost);
}

template<typename T>
template<typename TFirst, typename... TRest>
void Classifier<T>::EarlyRelease(TFirst& pointer, TRest& ...rest) {
	EarlyRelease(pointer);
	EarlyRelease(rest...);
}

template<typename T>
template<typename TFirst>
void Classifier<T>::EarlyRelease(TFirst& pointer) {
	if (option.useEarlyRelease) {
		cudaFree(pointer);
		pointer = nullptr;
	}
}
